File: ios/ExpoPropertiesModule.swift
import ExpoModulesCore

public class ExpoPropertiesModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoProperties")
    
    Function("setString") { (key: String, value: String) -> Void in
      UserDefaults.standard.set(value, forKey: key)
    }

    Function("getString") { (key: String) -> String? in
      return UserDefaults.standard.string(forKey: key) ?? nil
    }
    
    Function("setNumber") { (key: String, value: Double) -> Void in
      UserDefaults.standard.set(value, forKey: key)
    }
    
    Function("getNumber") { (key: String) -> Double in
      return UserDefaults.standard.double(forKey: key)
    }

    Function("setBoolean") { (key: String, value: Bool) -> Void in
      UserDefaults.standard.set(value, forKey: key)
    }
    
    Function("getBoolean") { (key: String) -> Bool in
      return UserDefaults.standard.bool(forKey: key)
    }
  }
}

File: android/src/main/java/expo/modules/properties/ExpoPropertiesModule.kt
package expo.modules.properties

import android.content.Context
import android.content.SharedPreferences
import androidx.core.os.bundleOf
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.types.Enumerable

class ExpoPropertiesModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoProperties")

    Function("setString") { key: String, value: String ->
      getPreferences().edit().putString(key, value).commit()
    }

    Function("getString") { key: String? ->
      return@Function getPreferences().getString(key, null)
    }

    Function("setNumber") { key: String, value: Double ->
      getPreferences().edit().putString(key, value.toString()).commit()
    }

    Function("getNumber") { key: String ->
      var s: String? = getPreferences().getString(key, "0")
      return@Function s?.toDouble()
    }

    Function("setBoolean") { key: String, value: Boolean ->
      getPreferences().edit().putString(key, value.toString()).commit()
    }

    Function("getBoolean") { key: String ->
      var s: String? = getPreferences().getString(key, "false")
      return@Function s?.toBoolean()
    }
  }

  private val context
  get() = requireNotNull(appContext.reactContext)

  private fun getPreferences(): SharedPreferences {
    return context.getSharedPreferences(context.packageName + ".settings", Context.MODE_PRIVATE)
  }
}